token-lending/sdk/src/state/reserve.rs

```diff
@@ -304,6 +310,49 @@ impl Reserve {
         })
     }
 
+    /// Calculate bonus as a percentage
+    /// the value will be in range [0, MAX_BONUS_PCT]
+    pub fn calculate_bonus(&self, obligation: &Obligation) -> Result<Decimal, ProgramError> {
+        if obligation.borrowed_value < obligation.unhealthy_borrow_value {
+            msg!("Obligation is healthy so a liquidation bonus can't be calculated");
+            return Err(LendingError::ObligationHealthy.into());
+        }
+
+        // could also return the average of liquidation bonus and max liquidation bonus here, but
+        // i don't think it matters
+        if obligation.unhealthy_borrow_value == obligation.super_unhealthy_borrow_value {
+            return Ok(Decimal::from_percent(self.config.liquidation_bonus));
+        }
+
+        // safety:
+        // - super_unhealthy_borrow value > unhealthy borrow value because we verify
+        // the ge condition in Reserve::unpack and then verify that they're not equal from check
+        // above
+        // - borrowed_value is >= unhealthy_borrow_value bc of the check above
+        // => weight is always between 0 and 1
+        let weight = min(
+            obligation
+                .borrowed_value
+                .try_sub(obligation.unhealthy_borrow_value)?
+                .try_div(
+                    obligation
+                        .super_unhealthy_borrow_value
+                        .try_sub(obligation.unhealthy_borrow_value)?,
```

Comment:
hmm. if super_unhealthy_borrow_value and unhealthy_borrow_value are really close together, then this calculation could overflow

Reply:
unwrap_or(Deimal::from(1))